### PR TITLE
FEATURE: possibility to define global Classload for ND4J

### PR DESCRIPTION
**Pull request by hosuaby**
This PR is related to https://github.com/eclipse/deeplearning4j/issues/8712. This issue is talking about the need to have a possibility to provide class loader for method `Class.forName`. Author mention only class `WordVectorSerializer` from `deeplearning-nlp`. But the problem is more general. We need to have possibility to chose classloader used by ND4J and Deeplearning4j.

This PR is a first part of the work. I added a new class `ND4JClassLoading` that gives possibility to used to provide his own classloader used by `Class.forName` and `ServiceLoader.load`. And I replaced all call to those methods by calls to this new class. I hope I forgot nothing.

There will be a second part of this work in anouther PR, concerning Dl4j. Both PR will be inderpendent.

### CODE REVIEW

**Comment by agibsonccc** [comment]
@hosuaby hey sorry about this. I'm just getting to some of these, do you mind updating? I understand if you can't, otherwise we can get this merged.

**Comment by agibsonccc** [comment]
Hey, sorry this is late. Could you modify it to use the nd4j class loader? Otherwise, this seems legit. Sorry for the late response!

**Comment by hosuaby** [comment]
Hello, @agibsonccc :)
Thanks for review. First, I made a rebase of branch and solved conflicts with current master.
I don't understood your demend. If need any modification, don't hesitate to ask. What exactly should I modify ?

**Comment by hosuaby** [comment]
Hello, @agibsonccc .

I think I undestood what you wanna say. My mistake, I have better to explain my plan.

This PR is the first part of work concerning https://github.com/eclipse/deeplearning4j/issues/8712. It provides ability to define classloader for ND4J. For DL4J I would like to have a separate class `DL4JClassLoading`, who will behaivour the same way, but for Dl4j packages.

The reasons I wanna to keep two classloaders separate: 
1) Nd4j and Dl4j are semantically two distinct projects
2) In my use case, I only need to redefine classloader for ND4j.
3) If I decide to rely on `ND4JClassLoading` from dl4j packages, it's will not be clear for user why should he define classloader for nd4j instead of dl4j.

So, my plan is, if this PR is approved, I make immediately the second one with `DL4JClassLoading` class.

Tell me what you think about.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4jBackend.java [hunk 1]
```diff
@@ -156,14 +157,12 @@ public static Nd4jBackend load() throws NoAvailableBackendException {
         String logInitProperty = System.getProperty(ND4JSystemProperties.LOG_INITIALIZATION, "true");
         boolean logInit = Boolean.parseBoolean(logInitProperty);
 
-        List<Nd4jBackend> backends = new ArrayList<>(1);
-        ServiceLoader<Nd4jBackend> loader = ServiceLoader.load(Nd4jBackend.class);
+        List<Nd4jBackend> backends = new ArrayList<>();
+        ServiceLoader<Nd4jBackend> loader = ND4JClassLoading.loadService(Nd4jBackend.class);
```
Comment (agibsonccc) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4jBackend.java:162]: @hosuaby you might want to also change the loadLibrary call in Nd4jBackend. Could you go through and look through all the uses of classloaders and make sure whatever you do is propagated so the classloader takes effect everywhere?

**Review by agibsonccc — CHANGES_REQUESTED** [review]

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4jBackend.java (hunk 1 above)
Comment (hosuaby) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/factory/Nd4jBackend.java:162]: Done. PR updated and rebased on master

**Comment by hosuaby** [comment]
Hello, @agibsonccc 
Any updates about this PR ?

**Review by agibsonccc — CHANGES_REQUESTED** [review]
Minor nit

**Comment by agibsonccc** [comment]
@hosuaby minor nit then I think we can merge this. Thanks!

**Comment by hosuaby** [comment]
@agibsonccc Ok. Do not hesitate to ask any change!

**Comment by agibsonccc** [comment]
@hosuaby  think you could get this over the line? I think we can merge this soon.

**Comment by hosuaby** [comment]
@agibsonccc done

**Review by agibsonccc — APPROVED** [review]

**Comment by agibsonccc** [comment]
Great thanks @hosuaby  I want to ping @saudet for 1 review here. @saudet  do you see any major issues here?

**Comment by saudet** [comment]
I didn't test it, but it looks fine.

**Comment by hosuaby** [comment]
@agibsonccc Thanks!
Now I will take care of #8712 that is very similar.

**Pull request merged**